<a href="https://colab.research.google.com/github/Fordalo/Data_SCIENCE_WORK/blob/main/decomposition_notes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Eigenvectors
In this demo, We will be discussing eigen decomposition and QR decomposition and why they are useful in data science practices. 


Lets start with eigen decomposition. 
 
*   matrix is singular if and only if one of the eigenvalues are zero.
*   Under Specific conditions, they can optomize quadratic expressions


          1.   largest eigen value = maximum of f(x)
          2.   smallest = minimum of f(x)



Here is the basic formula below:

X = $V$$\Lambda$ $V^{-1}$

Where X represents our original matrix, V represents all of the eigenvectors together, Λ = a diagonal matrix with all of of the eigenvalues. Lets build a quick martix below in python and try to extract the eigen values and vectors


In [12]:
import numpy as np
import pandas as pd
X = np.array([[10,4], [3,7]])

lambda1, V = np.linalg.eig(X)

print('Eigenvalues: {} \nEigenvectors: \n{}'.format(lambda1, V))


Eigenvalues: [12.27491722  4.72508278] 
Eigenvectors: 
[[ 0.86925207 -0.60422718]
 [ 0.49436913  0.79681209]]


**12.27491722**   would be the first characteristic of X

Lets see if we can satisfy our eigenvector equation "X = $V$$\Lambda$ $V^{-1}$" with python

In [13]:
vinvV = np.linalg.inv(V)
Lambda = np.diag(lambda1)
Xalt = V @ Lambda @ vinvV
Xalt

array([[10.,  4.],
       [ 3.,  7.]])

As you can see, we got back our original matrix above. Calculating the inverse of a matrix is always computationally expensive, so what if X was a symmetrical matrix? 